### PR TITLE
HIVE-26711 The very first REPL Load should make the Target Database read-only.

### PR DESCRIPTION
**Pull request by shreenidhiSaigaonkar**
Details :
* Basically first load(Boostrap load) set's the target db readonly.
* During incremental load it checks if target db is set to readonly, if not set then it makes target db readonly (basically to ensure backward compatibility)

Tests :
* Added unit tests to cover the same
* Tested manually in local environment

### What changes were proposed in this pull request?
This makes TARGET database readonly just after first bootstrap load.

### Why are the changes needed?
It ensures the TARGET database is always consistent and is only used for syncing changes from the SOURCE

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
* added unit tests from the same
* Also tested this in local environment

### CODE REVIEW

**Comment by shreenidhiSaigaonkar** [comment]
@pudidic could you please review the change

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 1]
```diff
@@ -420,9 +422,40 @@ a database ( directory )
         // Ignore if no file.
       }
     }
+
+    if(isReadOnlyHookRegistered()) {
```
Comment (pudidic) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:426]: Please follow coding convention `if(isReadOnlyHookRegistered()) {`

Comment (shreenidhiSaigaonkar) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java]: Done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplWithReadOnlyHook.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    overrides.put(MetastoreConf.ConfVars.EVENT_MESSAGE_FACTORY.getHiveName(),
+      GzipJSONMessageEncoder.class.getCanonicalName());
+
+    conf = new HiveConf(TestReplWithReadOnlyHook.class);
+    conf.set("hadoop.proxyuser." + Utils.getUGI().getShortUserName() + ".hosts", "*");
+
+    MiniDFSCluster miniDFSCluster =
+      new MiniDFSCluster.Builder(conf).numDataNodes(2).format(true).build();
+
+    Map<String, String> acidEnableConf = new HashMap<String, String>() {{
```
Comment (pudidic) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplWithReadOnlyHook.java:51]: It's my first time to see this double brace initialization in my Java experience. So I may be wrong about this. I did a small research, and some of references call it an anti-pattern. How about just repeating with HashMap calls or using ImmutableMap.Builder to chain the methods? Thanks.

1. https://www.baeldung.com/java-double-brace-initialization
2. https://blog.jooq.org/dont-be-clever-the-double-curly-braces-anti-pattern/

**Comment by pudidic** [comment]
Also please check the CI build. It may need a re-triggering with an empty commit. `git commit --allow-empty -m 'message'`.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplWithReadOnlyHook.java (hunk 2 above)
Comment (shreenidhiSaigaonkar) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplWithReadOnlyHook.java]: Done

**Comment by pudidic** [comment]
LGTM +1. I'll merge it.

**Pull request merged**